spec/models/have_a_degree_spec.rb

```diff
@@ -48,4 +48,64 @@
       end
     end
   end
+
+  describe "#set_degree_status" do
+    context "with degree_option degree" do
+      it "sets the correct degree_status_id before validation" do
+        have_a_degree.valid?
+        expect(have_a_degree.degree_status_id).to eq(HaveADegree::DEGREE_STATUS_OPTIONS[:yes])
+      end
+    end
+
+    context "with degree_option no" do
+      it "sets the correct degree_status_id before validation" do
+        no.valid?
+        expect(no.degree_status_id).to eq(HaveADegree::DEGREE_STATUS_OPTIONS[:no])
+      end
+    end
+
+    context "with degree_option studying" do
+      it "sets the correct degree_status_id before validation" do
+        studying.valid?
+        expect(studying.degree_status_id).to eq(HaveADegree::DEGREE_STATUS_OPTIONS[:studying])
+      end
+    end
+
+    context "with degree_option equivalent" do
+      it "sets the correct degree_status_id before validation" do
+        equivalent.valid?
+        expect(equivalent.degree_status_id).to eq(HaveADegree::DEGREE_STATUS_OPTIONS[:yes])
+      end
+    end
+  end
+
+  describe "#set degree type" do
+    context "with degree_option degree" do
+      it "sets the correct degree_type_id before validation" do
+        have_a_degree.valid?
+        expect(have_a_degree.degree_type_id).to eq(HaveADegree::DEGREE_TYPE[:degree])
+      end
+    end
+
+    context "with degree_option no" do
+      it "sets the correct degree_type_id before validation" do
+        no.valid?
```

Comment:
It may be worth overriding the setter for `degree_options` instead of doing the logic in `before_validation` hooks as it may catch us out in the future by reading the value before running the `.valid?` method

Reply:
Sorry, I don't quite follow ?

Reply:
So at the minute if you did something like:

```
model.degree_options = "degree"
```

You need to know to call `.valid?` before trying to read the `model.degree_status_id` or `model.degree_type_id`. It feels like something that will potentially trip us up in the future. Instead, you could override the setter of `degree_options`, something like this:

```
def degree_options=(value)
  super
  set_degree_status
  set_degree_type  
end
```

That way you can do this and get the correct values:

```
model.degree_options = "degree"
puts model.degree_status_id
puts model.degree_type_id
```

^ no longer needing the call to `valid?` in between

Reply:
👍